pysoa/server/logging.py

```diff
@@ -36,3 +38,151 @@ def get_logging_request_context(cls):
+        'card_verification_codes', 'card-verification-codes', 'cardVerificationCodes',
+        'card_verification_number', 'card-verification-number', 'cardVerificationNumber',
+        'card_verification_numbers', 'card-verification-numbers', 'cardVerificationNumbers',
+        'card_verification_value', 'card-verification-value', 'cardVerificationValue',
+        'card_verification_values', 'card-verification-values', 'cardVerificationValues',
+        'account_number', 'account-number', 'accountNumber',
+        'account_numbers', 'account-numbers', 'accountNumbers',
+        'bank_account', 'bank-account', 'bankAccount',
+        'bank_accounts', 'bank-accounts', 'bankAccounts',
+        'bank_account_number', 'bank-account-number', 'bankAccountNumber',
+        'bank_account_numbers', 'bank-account-numbers', 'bankAccountNumbers',
+        'pin', 'pin_code', 'pin-code', 'pinCode', 'pin_number', 'pin-number', 'pinNumber',
+        'pins', 'pin_codes', 'pin-codes', 'pinCodes', 'pin_numbers', 'pin-numbers', 'pinNumbers',
+        'personal_id_number', 'personal-id-number', 'personalIdNumber',
+        'personal_id_numbers', 'personal-id-numbers', 'personalIdNumbers',
+        'personal_identification_number', 'personal-identification-number', 'personalIdentificationNumber',
+        'personal_identification_numbers', 'personal-identification-numbers', 'personalIdentificationNumbers',
+    })
+
+    CENSOR_TYPES = six.string_types + six.integer_types
+
+    CENSORED_STRING = '**********'
+
+    def __init__(self, wrapped_dict):
+        """
+        Wraps a dict to censor its contents. The first time `repr` is called, it copies the dict, recursively
+        censors sensitive fields, caches the result, and returns the censored dict repr-ed. All future calls use
+        the cache.
+
+        :param wrapped_dict: The `dict` that should be censored
+        :type wrapped_dict: dict
+        """
+        if not isinstance(wrapped_dict, dict):
+            raise ValueError('wrapped_dict must be a dict')
+
+        self._wrapped_dict = wrapped_dict
+        self._dict_cache = None
+        self._repr_cache = None
+
+    def _get_repr_cache(self):
+        if not self._dict_cache:
+            self._dict_cache = self._copy_and_censor_dict(self._wrapped_dict)
+
+        return repr(self._dict_cache)
+
+    def __repr__(self):
+        if not self._repr_cache:
+            self._repr_cache = self._get_repr_cache()
+        return self._repr_cache
+
+    def __str__(self):
+        return self.__repr__()
+
+    def __bytes__(self):
+        # If this method is called, we must be in Python 3, which means __str__ must be returning a string
+        return self.__str__().encode('utf-8')
+
+    def __unicode__(self):
+        # If this method is called, we must be in Python 2, which means __str__ must be returning bytes
+        return self.__str__().decode('utf-8')
+
+    @classmethod
+    def _copy_and_censor_unknown_value(cls, v, should_censor_strings):
```

Comment:
at the point where it's used, can the value be any custom type?

Reply:
I'm not sure what you mean by this. Can you re-phrase?

Reply:
can `v` be anything but the basic types?

Reply:
Yes. If it's not one of the basic types, it won't be censored, but we only support serialization/deserialization of the basic types, anyway, so this isn't a problem.